// @ckeditor/ckeditor5-angular/bundles/ckeditor-ckeditor5-angular.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ckeditor/ckeditor5-watchdog/src/editorwatchdog'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@ckeditor/ckeditor5-angular', ['exports', '@angular/core', '@ckeditor/ckeditor5-watchdog/src/editorwatchdog', '@angular/forms', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.ckeditor = global.ckeditor || {}, global.ckeditor['ckeditor5-angular'] = {}), global.ng.core, global.EditorWatchdog, global.ng.forms, global.ng.common));
}(this, (function (exports, core, EditorWatchdog, forms, common) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var EditorWatchdog__default = /*#__PURE__*/_interopDefaultLegacy(EditorWatchdog);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
     * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
     */
    var ɵ0 = function (val, index) { return ('0' + (index).toString(16)).slice(-2); };
    // A copy of @ckeditor/ckeditor5-utils/src/uid.js
    // A hash table of hex numbers to avoid using toString() in uid() which is costly.
    // [ '00', '01', '02', ..., 'fe', 'ff' ]
    var HEX_NUMBERS = new Array(256).fill(0)
        .map(ɵ0);
    /**
     * Returns a unique id. The id starts with an "e" character and a randomly generated string of
     * 32 alphanumeric characters.
     *
     * **Note**: The characters the unique id is built from correspond to the hex number notation
     * (from "0" to "9", from "a" to "f"). In other words, each id corresponds to an "e" followed
     * by 16 8-bit numbers next to each other.
     *
     * @returns An unique id string.
     */
    function uid() {
        // Let's create some positive random 32bit integers first.
        //
        // 1. Math.random() is a float between 0 and 1.
        // 2. 0x100000000 is 2^32 = 4294967296.
        // 3. >>> 0 enforces integer (in JS all numbers are floating point).
        //
        // For instance:
        //		Math.random() * 0x100000000 = 3366450031.853859
        // but
        //		Math.random() * 0x100000000 >>> 0 = 3366450031.
        var r1 = Math.random() * 0x100000000 >>> 0;
        var r2 = Math.random() * 0x100000000 >>> 0;
        var r3 = Math.random() * 0x100000000 >>> 0;
        var r4 = Math.random() * 0x100000000 >>> 0;
        // Make sure that id does not start with number.
        return 'e' +
            HEX_NUMBERS[r1 >> 0 & 0xFF] +
            HEX_NUMBERS[r1 >> 8 & 0xFF] +
            HEX_NUMBERS[r1 >> 16 & 0xFF] +
            HEX_NUMBERS[r1 >> 24 & 0xFF] +
            HEX_NUMBERS[r2 >> 0 & 0xFF] +
            HEX_NUMBERS[r2 >> 8 & 0xFF] +
            HEX_NUMBERS[r2 >> 16 & 0xFF] +
            HEX_NUMBERS[r2 >> 24 & 0xFF] +
            HEX_NUMBERS[r3 >> 0 & 0xFF] +
            HEX_NUMBERS[r3 >> 8 & 0xFF] +
            HEX_NUMBERS[r3 >> 16 & 0xFF] +
            HEX_NUMBERS[r3 >> 24 & 0xFF] +
            HEX_NUMBERS[r4 >> 0 & 0xFF] +
            HEX_NUMBERS[r4 >> 8 & 0xFF] +
            HEX_NUMBERS[r4 >> 16 & 0xFF] +
            HEX_NUMBERS[r4 >> 24 & 0xFF];
    }

    var CKEditorComponent = /** @class */ (function () {
        function CKEditorComponent(elementRef, ngZone) {
            /**
             * The configuration of the editor.
             * See https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_editorconfig-EditorConfig.html
             * to learn more.
             */
            this.config = {};
            /**
             * The initial data of the editor. Useful when not using the ngModel.
             * See https://angular.io/api/forms/NgModel to learn more.
             */
            this.data = '';
            /**
             * Tag name of the editor component.
             *
             * The default tag is 'div'.
             */
            this.tagName = 'div';
            /**
             * Fires when the editor is ready. It corresponds with the `editor#ready`
             * https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_editor-Editor.html#event-ready
             * event.
             */
            this.ready = new core.EventEmitter();
            /**
             * Fires when the content of the editor has changed. It corresponds with the `editor.model.document#change`
             * https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_model_document-Document.html#event-change
             * event.
             */
            this.change = new core.EventEmitter();
            /**
             * Fires when the editing view of the editor is blurred. It corresponds with the `editor.editing.view.document#blur`
             * https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:blur
             * event.
             */
            this.blur = new core.EventEmitter();
            /**
             * Fires when the editing view of the editor is focused. It corresponds with the `editor.editing.view.document#focus`
             * https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:focus
             * event.
             */
            this.focus = new core.EventEmitter();
            /**
             * Fires when the editor component crashes.
             */
            this.error = new core.EventEmitter();
            /**
             * If the component is read–only before the editor instance is created, it remembers that state,
             * so the editor can become read–only once it is ready.
             */
            this.initiallyDisabled = false;
            /**
             * A lock flag preventing from calling the `cvaOnChange()` during setting editor data.
             */
            this.isEditorSettingData = false;
            this.id = uid();
            this.ngZone = ngZone;
            this.elementRef = elementRef;
        }
        Object.defineProperty(CKEditorComponent.prototype, "disabled", {
            get: function () {
                if (this.editorInstance) {
                    return this.editorInstance.isReadOnly;
                }
                return this.initiallyDisabled;
            },
            /**
             * When set `true`, the editor becomes read-only.
             * See https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_editor-Editor.html#member-isReadOnly
             * to learn more.
             */
            set: function (isDisabled) {
                this.setDisabledState(isDisabled);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CKEditorComponent.prototype, "editorInstance", {
            /**
             * The instance of the editor created by this component.
             */
            get: function () {
                var editorWatchdog = this.editorWatchdog;
                if (this.watchdog) {
                    // Temporarily use the `_watchdogs` internal map as the `getItem()` method throws
                    // an error when the item is not registered yet.
                    // See https://github.com/ckeditor/ckeditor5-angular/issues/177.
                    editorWatchdog = this.watchdog._watchdogs.get(this.id);
                }
                if (editorWatchdog) {
                    return editorWatchdog.editor;
                }
                return null;
            },
            enumerable: false,
            configurable: true
        });
        // Implementing the AfterViewInit interface.
        CKEditorComponent.prototype.ngAfterViewInit = function () {
            this.attachToWatchdog();
        };
        // Implementing the OnDestroy interface.
        CKEditorComponent.prototype.ngOnDestroy = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.watchdog) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.watchdog.remove(this.id)];
                        case 1:
                            _a.sent();
                            return [3 /*break*/, 4];
                        case 2:
                            if (!(this.editorWatchdog && this.editorWatchdog.editor)) return [3 /*break*/, 4];
                            return [4 /*yield*/, this.editorWatchdog.destroy()];
                        case 3:
                            _a.sent();
                            this.editorWatchdog = undefined;
                            _a.label = 4;
                        case 4: return [2 /*return*/];
                    }
                });
            });
        };
        // Implementing the ControlValueAccessor interface (only when binding to ngModel).
        CKEditorComponent.prototype.writeValue = function (value) {
            // This method is called with the `null` value when the form resets.
            // A component's responsibility is to restore to the initial state.
            if (value === null) {
                value = '';
            }
            // If already initialized.
            if (this.editorInstance) {
                // The lock mechanism prevents from calling `cvaOnChange()` during changing
                // the editor state. See #139
                this.isEditorSettingData = true;
                this.editorInstance.setData(value);
                this.isEditorSettingData = false;
            }
            // If not, wait for it to be ready; store the data.
            else {
                this.data = value;
                // If the editor element is already available, then update its content.
                // If the ngModel is used then the editor element should be updated directly here.
                if (this.editorElement) {
                    this.editorElement.innerHTML = this.data;
                }
            }
        };
        // Implementing the ControlValueAccessor interface (only when binding to ngModel).
        CKEditorComponent.prototype.registerOnChange = function (callback) {
            this.cvaOnChange = callback;
        };
        // Implementing the ControlValueAccessor interface (only when binding to ngModel).
        CKEditorComponent.prototype.registerOnTouched = function (callback) {
            this.cvaOnTouched = callback;
        };
        // Implementing the ControlValueAccessor interface (only when binding to ngModel).
        CKEditorComponent.prototype.setDisabledState = function (isDisabled) {
            // If already initialized.
            if (this.editorInstance) {
                this.editorInstance.isReadOnly = isDisabled;
            }
            // Store the state anyway to use it once the editor is created.
            this.initiallyDisabled = isDisabled;
        };
        /**
         * Creates the editor instance, sets initial editor data, then integrates
         * the editor with the Angular component. This method does not use the `editor.setData()`
         * because of the issue in the collaboration mode (#6).
         */
        CKEditorComponent.prototype.attachToWatchdog = function () {
            var _this = this;
            var creator = function (element, config) { return __awaiter(_this, void 0, void 0, function () {
                var _this = this;
                return __generator(this, function (_a) {
                    return [2 /*return*/, this.ngZone.runOutsideAngular(function () { return __awaiter(_this, void 0, void 0, function () {
                            var editor;
                            var _this = this;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        this.elementRef.nativeElement.appendChild(element);
                                        return [4 /*yield*/, this.editor.create(element, config)];
                                    case 1:
                                        editor = _a.sent();
                                        if (this.initiallyDisabled) {
                                            editor.isReadOnly = this.initiallyDisabled;
                                        }
                                        this.ngZone.run(function () {
                                            _this.ready.emit(editor);
                                        });
                                        this.setUpEditorEvents(editor);
                                        return [2 /*return*/, editor];
                                }
                            });
                        }); })];
                });
            }); };
            var destructor = function (editor) { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, editor.destroy()];
                        case 1:
                            _a.sent();
                            this.elementRef.nativeElement.removeChild(this.editorElement);
                            return [2 /*return*/];
                    }
                });
            }); };
            var emitError = function () {
                _this.ngZone.run(function () {
                    _this.error.emit();
                });
            };
            var element = document.createElement(this.tagName);
            var config = this.getConfig();
            this.editorElement = element;
            // Based on the presence of the watchdog decide how to initialize the editor.
            if (this.watchdog) {
                // When the context watchdog is passed add the new item to it based on the passed configuration.
                this.watchdog.add({
                    id: this.id,
                    type: 'editor',
                    creator: creator,
                    destructor: destructor,
                    sourceElementOrData: element,
                    config: config
                });
                this.watchdog.on('itemError', function (_, _a) {
                    var itemId = _a.itemId;
                    if (itemId === _this.id) {
                        emitError();
                    }
                });
            }
            else {
                // In the other case create the watchdog by hand to keep the editor running.
                var editorWatchdog = new EditorWatchdog__default['default'](this.editor);
                editorWatchdog.setCreator(creator);
                editorWatchdog.setDestructor(destructor);
                editorWatchdog.on('error', emitError);
                this.editorWatchdog = editorWatchdog;
                this.editorWatchdog.create(element, config);
            }
        };
        CKEditorComponent.prototype.getConfig = function () {
            if (this.data && this.config.initialData) {
                throw new Error('Editor data should be provided either using `config.initialData` or `data` properties.');
            }
            // Merge two possible ways of providing data into the `config.initialData` field.
            return Object.assign(Object.assign({}, this.config), { initialData: this.config.initialData || this.data || '' });
        };
        /**
         * Integrates the editor with the component by attaching related event listeners.
         */
        CKEditorComponent.prototype.setUpEditorEvents = function (editor) {
            var _this = this;
            var modelDocument = editor.model.document;
            var viewDocument = editor.editing.view.document;
            modelDocument.on('change:data', function (evt) {
                _this.ngZone.run(function () {
                    if (_this.cvaOnChange && !_this.isEditorSettingData) {
                        var data = editor.getData();
                        _this.cvaOnChange(data);
                    }
                    _this.change.emit({ event: evt, editor: editor });
                });
            });
            viewDocument.on('focus', function (evt) {
                _this.ngZone.run(function () {
                    _this.focus.emit({ event: evt, editor: editor });
                });
            });
            viewDocument.on('blur', function (evt) {
                _this.ngZone.run(function () {
                    if (_this.cvaOnTouched) {
                        _this.cvaOnTouched();
                    }
                    _this.blur.emit({ event: evt, editor: editor });
                });
            });
        };
        return CKEditorComponent;
    }());
    CKEditorComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'ckeditor',
                    template: '<ng-template></ng-template>',
                    // Integration with @angular/forms.
                    providers: [
                        {
                            provide: forms.NG_VALUE_ACCESSOR,
                            // eslint-disable-next-line @typescript-eslint/no-use-before-define
                            useExisting: core.forwardRef(function () { return CKEditorComponent; }),
                            multi: true
                        }
                    ]
                },] }
    ];
    CKEditorComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.NgZone }
    ]; };
    CKEditorComponent.propDecorators = {
        editor: [{ type: core.Input }],
        config: [{ type: core.Input }],
        data: [{ type: core.Input }],
        tagName: [{ type: core.Input }],
        watchdog: [{ type: core.Input }],
        disabled: [{ type: core.Input }],
        ready: [{ type: core.Output }],
        change: [{ type: core.Output }],
        blur: [{ type: core.Output }],
        focus: [{ type: core.Output }],
        error: [{ type: core.Output }]
    };

    /**
     * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
     * For licensing, see LICENSE.md.
     */
    var CKEditorModule = /** @class */ (function () {
        function CKEditorModule() {
        }
        return CKEditorModule;
    }());
    CKEditorModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [forms.FormsModule, common.CommonModule],
                    declarations: [CKEditorComponent],
                    exports: [CKEditorComponent]
                },] }
    ];

    /**
     * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
     * For licensing, see LICENSE.md.
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CKEditorComponent = CKEditorComponent;
    exports.CKEditorModule = CKEditorModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
